hivemind/utils/mpfuture.py

```diff
@@ -111,91 +103,171 @@ def _set_event_threadsafe(self):
             raise InvalidStateError(f"Can't set_result to a future that is {self._state} ({self._uid})")
+        elif os.getpid() == self._origin_pid:
+            MPFuture._active_futures.pop(self._uid, None)
+            self._set_event_if_necessary()
         else:
-            self._state_cache[self._state], self._result = base.FINISHED, result
-            self._send_update(UpdateType.RESULT, result)
+            self._send_update(MessageType.RESULT, result)
+        super().set_result(result)
 
     def set_exception(self, exception: Optional[BaseException]):
-        if os.getpid() == self._origin_pid:
-            super().set_exception(exception)
-            MPFuture._active_futures.pop(self._uid, None)
-        elif self._state in TERMINAL_STATES:
+        if self._state in TERMINAL_STATES:
             raise InvalidStateError(f"Can't set_exception to a future that is {self._state} ({self._uid})")
+        elif os.getpid() == self._origin_pid:
+            MPFuture._active_futures.pop(self._uid, None)
+            self._set_event_if_necessary()
         else:
-            self._state_cache[self._state], self._exception = base.FINISHED, exception
-            self._send_update(UpdateType.EXCEPTION, exception)
+            self._send_update(MessageType.EXCEPTION, exception)
+        super().set_exception(exception)
 
     def cancel(self) -> bool:
-        if os.getpid() == self._origin_pid:
-            MPFuture._active_futures.pop(self._uid, None)
-            return super().cancel()
-        elif self._state in [base.RUNNING, base.FINISHED]:
+        if self._state in [base.RUNNING, base.FINISHED]:
             return False
+        elif os.getpid() == self._origin_pid:
+            MPFuture._active_futures.pop(self._uid, None)
+            self._set_event_if_necessary()
         else:
-            self._state_cache[self._state] = base.CANCELLED
-            self._send_update(UpdateType.CANCEL)
-            return True
+            self._send_update(MessageType.CANCEL)
+        return super().cancel()
 
     def set_running_or_notify_cancel(self):
-        if self._state == base.PENDING:
-            self._state = base.RUNNING
-            return True
-        elif self._state == base.CANCELLED:
-            return False
-        else:
-            raise InvalidStateError(
-                f"Can't set_running_or_notify_cancel when future is in {self._state} ({self._uid})"
-            )
+        """if synchronize is set to False, this future will ignore any state changes from origin"""
+        self._synchronize_if_necessary()
+        try:
+            is_running = super().set_running_or_notify_cancel()
+            if is_running and os.getpid() != self._origin_pid:
+                self._send_update(MessageType.RUNNING)
+            return is_running
+        except RuntimeError as e:
+            raise InvalidStateError(str(e))
 
     def result(self, timeout: Optional[float] = None) -> ResultType:
```

Comment:
Can we replace the part from `elif self._state == base.CANCELLED:` with just `return super().result()`?

Reply:
good catch